### PR TITLE
[beta] v1 legacy roles MVP

### PR DESCRIPTION
**Pull request by jctanner**
#### What is this PR doing:

Issue: AAH-1812

#### Notes: 

**PR Author**: Add a QE reviewer ([exceptions](https://docs.engineering.redhat.com/display/AUTOHUB/Other+Team+Processes#OtherTeamProcesses-galaxy_ngrepo)); 
**Reviewers**: look for sound code, no [code smells](https://www.codegrip.tech/productivity/everything-you-need-to-know-about-code-smells/), docs & test coverage
**Merger**: When merging, include the Jira issue link in the squashed commit

### CODE REVIEW

File: galaxy_ng/app/api/v1/viewsets.py
Comment (awcrosby) [inline galaxy_ng/app/api/v1/viewsets.py:141]: is `platforms_autocomplete` handled below?

File: galaxy_ng/app/api/v1/viewsets.py
Comment (awcrosby) [inline galaxy_ng/app/api/v1/viewsets.py:286]: for when the number of tasks grows large, i wonder if doing an order by creation date could prevent iterating over a large number, since likely task lookups will be for recent tasks.

Comment (jctanner) [reply galaxy_ng/app/api/v1/viewsets.py]: done

File: galaxy_ng/app/api/v1/viewsets.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/viewsets.py:141]: I think I need to add platforms to the serializer first.

File: galaxy_ng/app/api/v1/tasks.py
Comment (awcrosby) [inline galaxy_ng/app/api/v1/tasks.py:101]: This would prevent use of the private/internally used methods.
```suggestion
        readme = markup_utils.get_readme_doc_file(checkout_path)
        if not readme:
            raise Exception("No role readme found.")
        readme_html = markup_utils.get_html(readme)
```

File: galaxy_ng/app/utils/galaxy.py
Comment (awcrosby) [inline galaxy_ng/app/utils/galaxy.py:30]: Doesn't seem `github_repo` is ever passed in, but maybe that is expected for future edits

File: galaxy_ng/app/api/v1/viewsets.py
Comment (newswangerd) [inline galaxy_ng/app/api/v1/viewsets.py:90]: You might be able to clean up all this manual filtering by using a filterset (example: https://github.com/ansible/galaxy_ng/blob/master/galaxy_ng/app/api/ui/viewsets/collection.py#L155)

File: galaxy_ng/app/api/v1/tasks.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/tasks.py]: done, ty

File: galaxy_ng/app/api/v1/viewsets.py
Comment (newswangerd) [inline galaxy_ng/app/api/v1/viewsets.py:82]: This viewset breaks all of the DRF conventions. It should be split into multiple viewsets (roles, role versions and role content etc) with one for each resource and stick to using the default DRF methods (get_object, list, get, etc) instead of defining new methods for get_content, get_roles etc.

Check out the EE apis for a good example of this:
- viewsets: https://github.com/ansible/galaxy_ng/blob/master/galaxy_ng/app/api/ui/viewsets/execution_environment.py
- urls: https://github.com/ansible/galaxy_ng/blob/master/galaxy_ng/app/api/ui/urls.py#L62

File: galaxy_ng/app/utils/galaxy.py
Comment (newswangerd) [inline galaxy_ng/app/utils/galaxy.py:26]: What is this used for? Copying info from galaxy?

File: galaxy_ng/app/api/v1/tasks.py
Comment (newswangerd) [inline galaxy_ng/app/api/v1/tasks.py:23]: @awcrosby can galaxy_importer handle role imports?

File: galaxy_ng/app/api/v1/viewsets.py
Comment (newswangerd) [inline galaxy_ng/app/api/v1/viewsets.py:353]: This should use RBAC.

File: galaxy_ng/app/api/v1/urls.py
Comment (newswangerd) [inline galaxy_ng/app/api/v1/urls.py:11]: I'm wondering if we should launch sync as a management command instead of an API call. Making this utility part of our public API seems unnecessary, since this should only be used for galaxy.ansible.com.

@awcrosby thoughts?

**Review by newswangerd — CHANGES_REQUESTED** [review]

File: galaxy_ng/app/api/v1/tasks.py
Comment (awcrosby) [reply galaxy_ng/app/api/v1/tasks.py:23]: galaxy_importer only handles collections... it has pieces that process a role inside a collection, but not a standalone role.

This method does use some the utils from galaxy_importer to process the role readme.

File: galaxy_ng/app/api/v1/urls.py
Comment (awcrosby) [reply galaxy_ng/app/api/v1/urls.py:11]: It might be helpful to have this callable with a task status viewable since this may be called daily, or on demand, and able to be run by multiple users of the beta site.

Also we can delete this endpoint after this is adopted on galaxy.ansible.com

**Review by awcrosby — COMMENTED** [review]
Provided comments. After all conversations are resolved, LGTM

File: galaxy_ng/app/utils/galaxy.py
Comment (jctanner) [reply galaxy_ng/app/utils/galaxy.py:30]: These will allow for future filtering of which roles to synchronize. For example we might want to sync everything from a specific github org/user or a github repo.

File: galaxy_ng/app/api/v1/urls.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/urls.py:11]: The endpoint won't be available if the featureflag/env-var isn't set.

File: galaxy_ng/app/api/v1/viewsets.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/viewsets.py]: done

File: galaxy_ng/app/api/v1/urls.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/urls.py]: done

File: galaxy_ng/app/api/v1/viewsets.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/viewsets.py]: done

File: galaxy_ng/app/api/v1/viewsets.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/viewsets.py]: done

File: galaxy_ng/app/api/v1/models.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/models.py]: I honestly don't know, but it's gone now.

File: galaxy_ng/app/api/v1/urls.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/urls.py]: fixed.

File: galaxy_ng/app/api/v1/viewsets.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/viewsets.py]: done

File: galaxy_ng/app/api/v1/serializers.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/serializers.py:32]: done

File: galaxy_ng/app/api/v1/viewsets.py
Comment (awcrosby) [reply galaxy_ng/app/api/v1/viewsets.py:74]: @newswangerd is it alright if we address the v1 RBAC edits in a subsequent PR that we work after this is merged?

Comment (newswangerd) [reply galaxy_ng/app/api/v1/viewsets.py:74]: I want RBAC to be consistent. If he handle permission checking in multiple places, it's going to get really hard to audit the security. It's not super hard to add the access policies for this, so I would like to just do it now rather than risking it falling through the cracks and creating security issues later.

Comment (awcrosby) [reply galaxy_ng/app/api/v1/viewsets.py:74]: I was thinking our `DEPLOYMENT_MODE=community` deployment from `master` progress may be blocked by this PR not being merged, but at least right now it is not, :+1:

File: galaxy_ng/app/api/v1/viewsets.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/viewsets.py]: fixed.

File: galaxy_ng/app/api/v1/viewsets.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/viewsets.py]: fixed.

File: galaxy_ng/app/api/v1/viewsets.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/viewsets.py]: done.

File: galaxy_ng/app/api/v1/viewsets.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/viewsets.py]: done

File: galaxy_ng/app/api/v1/models.py
Comment (jctanner) [reply galaxy_ng/app/api/v1/models.py:106]: done, but only with users, not groups.

File: galaxy_ng/app/api/v1/viewsets/users.py
Comment (newswangerd) [inline galaxy_ng/app/api/v1/viewsets/users.py:57]: the builtin retrieve should work here.

File: galaxy_ng/app/api/v1/viewsets/tasks.py
Comment (newswangerd) [inline galaxy_ng/app/api/v1/viewsets/tasks.py:62]: This system is going to have tens of thousands of tasks in it. There has to be a more efficient way to find the task than just iterating over each one. This will make it super easy to DoS the system by simply spamming api requests to this api endpoint with an task ID that doesn't exist

**Review by newswangerd — CHANGES_REQUESTED** [review]
General notes:
- There's a lot of misc `print` statements that need to be removed.

File: galaxy_ng/app/access_control/statements/legacy.py
Comment (jctanner) [reply galaxy_ng/app/access_control/statements/legacy.py:15]: The galaxy docs have been telling users for multiple years that the [legacy] namespace is auto-created for them:

https://galaxy.ansible.com/docs/contributing/importing.html#importing-roles

> The first time you logged in, Galaxy created a namespace matching your GitHub namespace, so you should see at least one namespace.

File: galaxy_ng/app/api/v1/viewsets/roles.py
Comment (newswangerd) [inline galaxy_ng/app/api/v1/viewsets/roles.py:61]: I think the default DRF method should work for this

File: galaxy_ng/app/api/v1/viewsets/roles.py
Comment (newswangerd) [inline galaxy_ng/app/api/v1/viewsets/roles.py:50]: do you need to order this in 3 places?

**Review by newswangerd — APPROVED** [review]
A couple of minor nitpicks, no deal breakers.

Before this goes into production I do want to see:
1. A proper RBAC implementation for legacy namespace
2. Replacement for the `username==legacy_namespace.name` logic to determine if a user owns a namespace
3. Some kind of system to migrate the existing users from galaxy into galaxy_ng so we don't have to rely on the user's github username matching one of the namespaces

**Pull request merged**
